### PR TITLE
Bump rand to 0.8 (and everything that touches it)

### PR DESCRIPTION
**Pull request by dpc**
The goal of this change is to takes us off old `rand` dependency, but to accomplish that it is necessary to upgrade lots of other dependencies (mostly whole bitcoin ecosystem to `bitcoin = 0.29.1`), and fix all the code afterwards.

I also had to rebase our forked version of rust-secp256k1-zkp: https://github.com/fedimint/rust-secp256k1-zkp/pull/1

### CODE REVIEW

**Comment by elsirion** [comment]
Oh wow, `rust-bitcoin` upgraded `rand` :tada: if it's only a question of getting `bitcoincore-rpc` upgraded let's fork it till upstream accepts the change?

**Comment by justinmoon** [comment]
Looks like `bitcoincore_rpc` merged this and it was released. The commit shows up in their [`0.16.0` git tag](https://github.com/rust-bitcoin/rust-bitcoincore-rpc/commits/v0.16.0). (is there a better way to see if a given commit has been released?

File: modules/fedimint-wallet/src/lib.rs [hunk 1]
```diff
@@ -1089,7 +1090,7 @@ impl<'a> StatelessWallet<'a> {
                 };
 
                 secret_key
-                    .add_assign(&tweak[..])
+                    .add_tweak(&Scalar::from_be_bytes(tweak).expect("can't fail"))
```
Comment (dpc) [inline modules/fedimint-wallet/src/lib.rs:1093]: le or be ?

File: modules/fedimint-wallet/src/tweakable.rs [hunk 2]
```diff
@@ -21,7 +21,8 @@ impl Tweakable for secp256k1::PublicKey {
         let tweak = Hmac::from_engine(hasher).into_inner();
 
         let mut key = *self;
-        key.add_exp_assign(secp, &tweak[..])
+        // TODO: is this supposed to be be or le?
+        key.add_exp_tweak(secp, &Scalar::from_be_bytes(tweak).expect("can't fail"))
```
Comment (dpc) [inline modules/fedimint-wallet/src/tweakable.rs:25]: le or be?

**Comment by dpc** [comment]
@justinmoon What should we do with https://github.com/fedimint/rust-secp256k1-zkp/pull/1 ?

Can anyone review the "Miniscript-specific fixes" commit specifically, as I'm not sure if I understand everything there. Also - I don't know if `Scalars` from bytes should be BE or LE.

File: modules/fedimint-wallet/src/tweakable.rs (hunk 2 above)
Comment (elsirion) [reply modules/fedimint-wallet/src/tweakable.rs]: Whatever we use otherwise, it doesn't make any difference afaiu.

**Review by jkitman — DISMISSED** [review]
After talking with @elsirion it's probably best we update `rand` in our libraries first.

Merging in DKG will update rand for threshold_crypto and hbbft (hope to do that by tomorrow).

For secp256k1-zkp we'll probably need to update our fork.

**Comment by dpc** [comment]
> After talking with @elsirion it's probably best we update `rand` in our libraries first.

While even better, it doesn't seem necessary to do in the same PR. Any other reasons?

> For secp256k1-zkp we'll probably need to update our fork.

This one is necessary and please verify the rebase https://github.com/fedimint/rust-secp256k1-zkp/pull/1 . I'll ask Justin to push as another branch in our repo today.

**Comment by jkitman** [comment]
> While even better, it doesn't seem necessary to do in the same PR. Any other reasons?

I'll let @elsirion answer, it was his recommendation.

**Comment by dpc** [comment]
I see #600 will update the crypto binaries. If you want to land it before the rand bump because it's more important that's fine with me.

**Comment by elsirion** [comment]
This one seems a tiny bit less ready, so merging #600 first. 

> This one is necessary and please verify the rebase https://github.com/fedimint/rust-secp256k1-zkp/pull/1 . I'll ask Justin to push as another branch in our repo today.

After this we should try to get this one in soon too. Feel free to merge each other PRs btw @jkitman @dpc :smiley:

**Comment by jkitman** [comment]
@dpc thanks for letting me get #600 in first, that was a painful one to rebase.

**Comment by dpc** [comment]
> I think now we should be able to remove the `Rand07Compat` wrapper and use `OsRng` everywhere, right? Will be much cleaner that way.

I hope so, yes.

**Review by justinmoon — APPROVED** [review]
LGTM

**Pull request merged**